### DISCUSSION TITLE
Differentiate between 'not set' and 'null' in UPDATE.

### DISCUSSION THREAD

**Discussion by webertim** [pocketbase/pocketbase#1294] [Q&A] [answered]
Is there any way to differentiate between `'not set'` and `null` in the API rules for an UPDATE action?

The use case for this is the following:
I want users to create entries in the database which they can only partially edit. For example they can create an object with `field1` and `field2`. After the creation the server edits `field2` according to some rules. After that the user should only be allowed to update `field1`.

First I tried to archive this by setting the following rule for the update action:
@request.data.field2 = null

There is a problem though. The user is now allowed to send either
```json
{
}
```
or 
```json
{
   "field2": null
}
```

The problem with the second request is, that the field value becomes `null`. A workaround is the following rule:
@request.data.field2 = field2

But therefore the client needs to do a new read request before every update since the server is editing field2 in the background.

A simple solution would be a value like `unset` that you could check in the update action rule. Is there anything like that?
If not I would be happy to submit a merge request for it.

**Comment by ganigeorgiev** [comment, marked as answer]
> A simple solution would be a value like unset that you could check in the update action rule. Is there anything like that?

No, this is not possible at the moment.

Technically, since PocketBase v0.3 we no longer support `null` as field value (_*with the exception of the `json` type field_) and internally it is always casted to the zero-default of the field type.

I'm not sure what would be the best way to define a missing value in a rule.
PR/suggestions are welcomed but please note that there will be changes to the filters handling in relation to the ongoing multi-match support and it may be better to wait for v0.11.

**Reply by webertim** [reply]
Thank you for the quick response!

I found another simple workaround (which does not work for every use case though). I keep the first rule I mentioned:
`@request.data.field2 = null`

And specify that the field as `Nonempty`

This makes the second request I mentioned impossible and therefore makes `field2` immutable.

**Reply by ganigeorgiev** [reply]
@webertim Yes, if the field is "Nonempty" this could be one way to do it but you may need to combine the rule with a direct match if you don't want to allow changes, aka.:
`(@request.data.field2 = null || @request.data.field2 = field2)`

**Reply by ganigeorgiev** [reply]
@webertim Since I'm planning to add field modifiers anyway, I'll consider with the next release to define a special `.isset` suffix for the `@request.*` filter fields that will resolve to `TRUE` only if the field is set in the request body (including `null`), aka. the final expression could look something like this:

```js
// this will ensure that the client didn't send "someField" with the request
@request.data.someField.isset = false
```

The modifier must be part of the operand because the field resolver parses the fields individually, aka. it doesn't know what is on the other side of the operator before building the final query or constraint.

I'm still not sure though about the `isset` naming, so feel free to let me know if you have other ideas.
